test/e2e/virtual_machine_test.go

```diff
@@ -0,0 +1,362 @@
+	}
+	t.Cleanup(func() {
+		if DestroyAfter {
+			_, err := server.NetworkSecurityGroupService.Delete(ctx, &resourcespb.DeleteNetworkSecurityGroupRequest{ResourceId: nsg.CommonParameters.ResourceId})
+			if err != nil {
+				logGrpcErrorDetails(t, err)
+				t.Logf("unable to delete resource %+v", err)
+			}
+		}
+	})
+
+	pubKey, privKey, err := makeSSHKeyPair()
+	fmt.Println(privKey)
+	if err != nil {
+		logGrpcErrorDetails(t, err)
+		t.Fatalf("unable to create ssh key: %+v", err)
+	}
+
+	var vmSize commonpb.VmSize_Enum
+	if cloud == commonpb.CloudProvider_AWS {
+		vmSize = commonpb.VmSize_MICRO
+	} else if cloud == commonpb.CloudProvider_AZURE {
+		vmSize = commonpb.VmSize_LARGE
+	}
+
+	createVmRequest := &resourcespb.CreateVirtualMachineRequest{Resource: &resourcespb.VirtualMachineArgs{
+		CommonParameters: &commonpb.ResourceCommonArgs{
+			Location:      commonpb.Location_EU_WEST_1,
+			CloudProvider: cloud,
+		},
+		Name:                    "vm-test-vm",
+		NetworkSecurityGroupIds: []string{nsg.CommonParameters.ResourceId},
+		VmSize:                  vmSize,
+		UserDataBase64: base64.StdEncoding.EncodeToString([]byte(`#!/bin/bash
+sudo echo "hello world" > /tmp/test.txt`)),
+		SubnetId:         publicSubnet.CommonParameters.ResourceId,
+		PublicSshKey:     pubKey,
+		GeneratePublicIp: true,
+		ImageReference: &resourcespb.ImageReference{
+			Os:      resourcespb.ImageReference_UBUNTU,
+			Version: "18.04",
+		},
+	}}
+	vm, err := server.VirtualMachineService.Create(ctx, createVmRequest)
+	if err != nil {
+		logGrpcErrorDetails(t, err)
+		t.Fatalf("unable to create virtual machine: %+v", err)
+	}
+	t.Cleanup(func() {
+		if DestroyAfter {
+			_, err := server.VirtualMachineService.Delete(ctx, &resourcespb.DeleteVirtualMachineRequest{ResourceId: vm.CommonParameters.ResourceId})
+			if err != nil {
+				logGrpcErrorDetails(t, err)
+				t.Logf("unable to delete resource %+v", err)
+			}
+		}
+	})
+
+	var username string
+	if cloud == commonpb.CloudProvider_AWS {
+		username = "ubuntu"
+	} else if cloud == commonpb.CloudProvider_AZURE {
+		username = "adminuser"
+	}
+
+	signer, err := signerFromPem([]byte(privKey), []byte(""))
+	if err != nil {
+		t.Fatal(fmt.Errorf("error setting up cert"))
+	}
+	config := &ssh.ClientConfig{
+		User: username,
+		Auth: []ssh.AuthMethod{
+			ssh.PublicKeys(signer),
+		},
+		HostKeyCallback: func(hostname string, remote net.Addr, key ssh.PublicKey) error {
+			return nil
+		},
+	}
+
+	fmt.Println(vm.GetPublicIp())
+
+	time.Sleep(3 * time.Minute)
+
+	// connect ot ssh server
+	conn, err := ssh.Dial("tcp", vm.GetPublicIp()+":22", config)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer conn.Close()
+
+	session, err := conn.NewSession()
+	if err != nil {
+		t.Fatal(fmt.Errorf("error creating ssh session"))
+	}
+	defer session.Close()
+
+	// run command and capture stdout/stderr
+	output, err := session.CombinedOutput("sudo cat /tmp/test.txt")
+	assert.Equal(t, string(output), "hello world\n")
```

Comment:
this lib wants the expected before the actual, which is odd if you ask me..

assert.Equal(t, "hello world\n", string(output))